File: host/tests/test_protocol.py
"""Protocol codec tests, including the known-answer vectors shared with the
Rust implementation (see docs/protocol.md and helic-proto)."""

import unittest

from helic_daq import protocol
from helic_daq.protocol import MsgType, ProtocolError, StreamHeader


class TestCrc16(unittest.TestCase):
    def test_known_answers(self):
        # Same vectors as helic-proto/src/crc.rs.
        self.assertEqual(protocol.crc16(b"123456789"), 0x29B1)
        self.assertEqual(protocol.crc16(b""), 0xFFFF)
        self.assertEqual(protocol.crc16(bytes([0x00])), 0xE1F0)
        self.assertEqual(protocol.crc16(bytes([0x0A, 0x01, 0x00, 0x00])), 0xDB5B)


class TestFrame(unittest.TestCase):
    def test_known_answer_frame(self):
        # Status request, seq 1, empty payload — same vector as helic-proto.
        frame = protocol.encode_frame(MsgType.STATUS, 1)
        self.assertEqual(frame, bytes([0x48, 0x4C, 0x0A, 0x01, 0x00, 0x00, 0x5B, 0xDB]))

    def test_known_answer_discovery_requests(self):
        self.assertEqual(
            protocol.encode_frame(MsgType.GET_PARAMS, 1),
            bytes.fromhex("48 4C 01 01 00 00 44 C5"),
        )
        self.assertEqual(
            protocol.encode_frame(MsgType.GET_SOURCES, 1),
            bytes.fromhex("48 4C 02 01 00 00 98 5E"),
        )

    def test_known_answer_v2_frames(self):
        block = protocol.encode_set_block(12, 0x01020304, b"\xaa\xbb")
        self.assertEqual(
            protocol.encode_frame(MsgType.SET_BLOCK, 2, block),
            bytes.fromhex("48 4C 05 02 08 00 0C 00 04 03 02 01 AA BB 39 A7"),
        )
        commit = protocol.encode_commit(12, 0x01020304)
        self.assertEqual(
            protocol.encode_frame(MsgType.COMMIT, 3, commit),
            bytes.fromhex("48 4C 06 03 06 00 0C 00 04 03 02 01 08 D1"),
        )
        status = bytes.fromhex("02 11 00 0D 00 00 FA 45 10 A4 00 00")
        self.assertEqual(
            protocol.encode_frame(MsgType.STATUS, 1, status),
            bytes.fromhex(
                "48 4C 0A 01 0C 00 02 11 00 0D 00 00 FA 45 10 A4 00 00 03 09"
            ),
        )

    def test_round_trip(self):
        frame = protocol.encode_frame(MsgType.GET_PAR, 7, b"\x01\x00\x02\x00")
        msg_type, seq, payload = protocol.decode_frame(frame)
        self.assertEqual(msg_type, MsgType.GET_PAR)
        self.assertEqual(seq, 7)
        self.assertEqual(payload, b"\x01\x00\x02\x00")

    def test_corrupt_crc_rejected(self):
        frame = bytearray(protocol.encode_frame(3, 0, b"\x2a"))
        frame[protocol.HEADER_LEN] ^= 0xFF
        with self.assertRaises(ProtocolError):
            protocol.decode_frame(bytes(frame))

    def test_bad_magic_rejected(self):
        frame = bytearray(protocol.encode_frame(3, 0))
        frame[0] = 0
        with self.assertRaises(ProtocolError):
            protocol.decode_frame(bytes(frame))

    def test_oversize_payload_rejected(self):
        with self.assertRaises(ProtocolError):
            protocol.encode_frame(3, 0, bytes(protocol.MAX_PAYLOAD + 1))


class TestPayload(unittest.TestCase):
    def test_known_answer_discovery_entries(self):
        params = protocol.decode_params(b"freq\0f\x01\x00\x01")
        self.assertEqual(params, [("freq", "f", 1, True)])
        sources = protocol.decode_sources(b"adc0\0V\0laser\0mm\0")
        self.assertEqual(sources, [("adc0", "V"), ("laser", "mm")])

    def test_known_answer_block_payloads(self):
        self.assertEqual(
            protocol.encode_set_block(12, 0x01020304, b"\xaa\xbb"),
            b"\x0c\x00\x04\x03\x02\x01\xaa\xbb",
        )
        self.assertEqual(
            protocol.encode_commit(12, 0x01020304),
            b"\x0c\x00\x04\x03\x02\x01",
        )

    def test_malformed_discovery_rejected(self):
        with self.assertRaises(ProtocolError):
            protocol.decode_params(b"freq\0f")
        with self.assertRaises(ProtocolError):
            protocol.decode_sources(b"adc0\0V")


class TestBeacon(unittest.TestCase):
    def test_known_request_and_response_round_trip(self):
        self.assertEqual(protocol.BEACON_REQUEST, bytes.fromhex("48 4c 01"))
        beacon = protocol.BeaconResponse(
            2, 2350, bytes.fromhex("02 48 4c 00 00 01"), "cbc-rig", "helic-daq sim"
        )
        encoded = protocol.encode_beacon_response(beacon)
        self.assertEqual(
            encoded,
            bytes.fromhex(
                "48 4c 02 02 2e 09 02 48 4c 00 00 01 "
                "63 62 63 2d 72 69 67 00 00 00 00 00 00 00 00 00 "
                "68 65 6c 69 63 2d 64 61 71 20 73 69 6d 00 00 00"
            ),
        )
        self.assertEqual(protocol.decode_beacon_response(encoded), beacon)

    def test_malformed_response_is_rejected(self):
        with self.assertRaises(ProtocolError):
            protocol.decode_beacon_response(b"")


class TestStreamHeader(unittest.TestCase):
    def test_round_trip(self):
        h = StreamHeader(
            n_sources=12, seq=123456, first_index=42, dropped=3, decimation=2, n_records=28
        )
        buf = protocol.encode_stream_header(h)
        self.assertEqual(len(buf), protocol.STREAM_HEADER_LEN)
        self.assertEqual(protocol.decode_stream_header(buf), h)

    def test_bad_magic_rejected(self):
        h = StreamHeader(1, 0, 0, 0, 1, 0)
        buf = bytearray(protocol.encode_stream_header(h))
        buf[0] = 0
        with self.assertRaises(ProtocolError):
            protocol.decode_stream_header(bytes(buf))


if __name__ == "__main__":
    unittest.main()

File: host-matlab/tests/TestProtocol.m
% Known-answer and validation tests for the MATLAB protocol codec.
classdef TestProtocol < matlab.unittest.TestCase
    methods (Test)
        function crcKnownAnswers(testCase)
            testCase.verifyEqual(helicdaq.Protocol.crc16(uint8('123456789')), ...
                uint16(hex2dec('29B1')));
            testCase.verifyEqual(helicdaq.Protocol.crc16(uint8([])), ...
                uint16(hex2dec('FFFF')));
            testCase.verifyEqual(helicdaq.Protocol.crc16(uint8(0)), ...
                uint16(hex2dec('E1F0')));
        end

        function frameKnownAnswers(testCase)
            testCase.verifyEqual( ...
                helicdaq.Protocol.encodeFrame(helicdaq.Protocol.GET_PARAMS, 1), ...
                TestProtocol.hexBytes('48 4C 01 01 00 00 44 C5'));
            testCase.verifyEqual( ...
                helicdaq.Protocol.encodeFrame(helicdaq.Protocol.GET_SOURCES, 1), ...
                TestProtocol.hexBytes('48 4C 02 01 00 00 98 5E'));
            block = helicdaq.Protocol.encodeSetBlock(12, hex2dec('01020304'), ...
                uint8([hex2dec('AA'), hex2dec('BB')]));
            testCase.verifyEqual( ...
                helicdaq.Protocol.encodeFrame(helicdaq.Protocol.SET_BLOCK, 2, block), ...
                TestProtocol.hexBytes( ...
                '48 4C 05 02 08 00 0C 00 04 03 02 01 AA BB 39 A7'));
            commit = helicdaq.Protocol.encodeCommit(12, hex2dec('01020304'));
            testCase.verifyEqual( ...
                helicdaq.Protocol.encodeFrame(helicdaq.Protocol.COMMIT, 3, commit), ...
                TestProtocol.hexBytes( ...
                '48 4C 06 03 06 00 0C 00 04 03 02 01 08 D1'));
        end

        function frameRoundTripAndValidation(testCase)
            frame = helicdaq.Protocol.encodeFrame(helicdaq.Protocol.GET_PAR, 7, ...
                uint8([1, 0, 2, 0]));
            decoded = helicdaq.Protocol.decodeFrame(frame);
            testCase.verifyEqual(decoded.MessageType, helicdaq.Protocol.GET_PAR);
            testCase.verifyEqual(decoded.Sequence, uint8(7));
            testCase.verifyEqual(decoded.Payload, uint8([1, 0, 2, 0]));
            frame(7) = bitxor(frame(7), uint8(255));
            testCase.verifyError(@() helicdaq.Protocol.decodeFrame(frame), ...
                'helicdaq:ProtocolError');
            testCase.verifyError(@() helicdaq.Protocol.encodeFrame(3, 0, ...
                zeros(1, helicdaq.Protocol.MAX_PAYLOAD + 1, 'uint8')), ...
                'helicdaq:ProtocolError');
        end

        function discoveryEntries(testCase)
            parameters = helicdaq.Protocol.decodeParams( ...
                uint8([uint8('freq'), 0, uint8('f'), 1, 0, 1]));
            testCase.verifyEqual(parameters.Name, "freq");
            testCase.verifyEqual(parameters.TypeCode, "f");
            testCase.verifyEqual(parameters.Count, uint16(1));
            testCase.verifyTrue(parameters.Writable);
            sources = helicdaq.Protocol.decodeSources( ...
                uint8([uint8('adc0'), 0, uint8('V'), 0, ...
                uint8('laser'), 0, uint8('mm'), 0]));
            testCase.verifyEqual(sources.Name, ["adc0"; "laser"]);
            testCase.verifyEqual(sources.Unit, ["V"; "mm"]);
            testCase.verifyError(@() helicdaq.Protocol.decodeSources( ...
                uint8([uint8('adc0'), 0, uint8('V')])), ...
                'helicdaq:ProtocolError');
        end

        function parameterTypesRoundTrip(testCase)
            cases = {
                'B', uint8(200);
                'b', int8(-100);
                'H', uint16(50000);
                'h', int16(-20000);
                'I', uint32(4000000000);
                'i', int32(-2000000000);
                'f', single(1.25)
                };
            for row = 1:size(cases, 1)
                code = cases{row, 1};
                value = cases{row, 2};
                encoded = helicdaq.Protocol.packParameter(code, 1, value);
                testCase.verifyEqual( ...
                    helicdaq.Protocol.unpackParameter(code, 1, encoded), value);
            end
            encoded = helicdaq.Protocol.packParameter('c', 8, "cbc");
            testCase.verifyEqual( ...
                helicdaq.Protocol.unpackParameter('c', 8, encoded), "cbc");
        end

        function beaconKnownAnswer(testCase)
            response = struct('Version', uint8(2), 'ControlPort', uint16(2350), ...
                'Mac', TestProtocol.hexBytes('02 48 4C 00 00 01'), ...
                'Experiment', "cbc-rig", 'Firmware', "helic-daq sim");
            encoded = helicdaq.Protocol.encodeBeaconResponse(response);
            testCase.verifyEqual(encoded, TestProtocol.hexBytes( ...
                ['48 4C 02 02 2E 09 02 48 4C 00 00 01 ' ...
                '63 62 63 2D 72 69 67 00 00 00 00 00 00 00 00 00 ' ...
                '68 65 6C 69 63 2D 64 61 71 20 73 69 6D 00 00 00']));
            decoded = helicdaq.Protocol.decodeBeaconResponse(encoded);
            testCase.verifyEqual(decoded.Experiment, "cbc-rig");
            testCase.verifyEqual(decoded.Firmware, "helic-daq sim");
            testCase.verifyEqual(decoded.Mac, response.Mac);
        end

        function streamHeaderRoundTrip(testCase)
            header = struct('NSources', uint8(12), 'Sequence', uint32(123456), ...
                'FirstIndex', uint32(42), 'Dropped', uint32(3), ...
                'Decimation', uint16(2), 'NRecords', uint16(28));
            decoded = helicdaq.Protocol.decodeStreamHeader( ...
                helicdaq.Protocol.encodeStreamHeader(header));
            testCase.verifyEqual(decoded, header);
        end
    end

    methods (Static, Access = private)
        function bytes = hexBytes(text)
            compact = regexprep(text, '\s', '');
            bytes = uint8(sscanf(compact, '%2x').');
        end
    end
end
